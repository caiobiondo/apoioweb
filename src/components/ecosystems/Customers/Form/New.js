export { default } from './organisms/New';
