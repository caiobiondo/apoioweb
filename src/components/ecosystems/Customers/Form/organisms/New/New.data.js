import gql from 'graphql-tag';

export const CreateCustomerMutation = gql`
  mutation CreateCustomer($input: CreateCustomerInput!) {
    createCustomer(input: $input) {
      customer {
        id
        name
        nickname
      }
    }
  }
`;
