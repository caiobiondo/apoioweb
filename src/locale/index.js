import 'url-search-params-polyfill';
import { addLocaleData } from 'react-intl';
import IntlMessageFormat from 'intl-messageformat';

import en from 'react-intl/locale-data/en';
import pt from 'react-intl/locale-data/pt';

const currentLocale = () => {
  const params = new URLSearchParams(document.location.search.substring(1));
  return params.get('lang') || 'pt-BR';
};

addLocaleData([...en, ...pt]);
export const locale = currentLocale();
export const messages = require('./messages').default;
export const flattenMessages = require('./utils').flattenMessages;

export const translate = (messageId, values = {}) => {
  const formatter = new IntlMessageFormat(messages[locale][messageId], locale);
  return formatter.format(values);
};
