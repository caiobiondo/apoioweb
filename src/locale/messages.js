/* eslint-enable sort-keys */
import myScoreMessages from './messages/my_score';
import orderMessages from './messages/order';
import customerMessages from './messages/customer';
import genericMessages from './messages/generic';
import customersListMessages from './messages/customers_list';
import formValidationMessages from './messages/form_validation';

export default {
  'en-US': {
    ...customerMessages['en-US'],
    ...customersListMessages['en-US'],
    ...formValidationMessages['en-US'],
    ...genericMessages['en-US'],
    ...myScoreMessages['en-US'],
    ...orderMessages['en-US'],
  },
  'pt-BR': {
    ...customerMessages['pt-BR'],
    ...customersListMessages['pt-BR'],
    ...formValidationMessages['pt-BR'],
    ...genericMessages['pt-BR'],
    ...myScoreMessages['pt-BR'],
    ...orderMessages['pt-BR'],
  },
};
/* eslint-disable sort-keys */
