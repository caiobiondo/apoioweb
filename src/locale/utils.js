export function flattenMessages(nestedMessages, prefix = '') {
  return Object.keys(nestedMessages).reduce((messages, key) => {
    const value = nestedMessages[key];
    const prefixedKey = prefix ? `${prefix}.${key}` : key;
    if (typeof value === 'string') {
      messages[prefixedKey] = value;
    } else {
      Object.assign(messages, flattenMessages(value, prefixedKey));
    }
    return messages;
  }, {});
}

export function formatDate(isoDate, intl, defaultValue = null) {
  if (!isoDate) return defaultValue;

  return intl.formatDate(new Date(isoDate), { day: 'numeric', month: 'numeric', year: 'numeric' });
}

export function formatTime(isoDate, intl, defaultValue = null) {
  if (!isoDate) return defaultValue;

  return intl.formatTime(new Date(isoDate));
}

export function formatCurrency(price, intl, defaultValue = null) {
  if (!price) return defaultValue;

  return intl.formatNumber(price, { currency: 'BRL', style: 'currency' });
}
