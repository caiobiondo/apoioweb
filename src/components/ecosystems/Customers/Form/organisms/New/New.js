import CustomerForm from '../Form';
import { withRouter } from 'react-router-dom';
import { CreateCustomerMutation } from './New.data';
import { graphql } from 'react-apollo';
import { withFormik } from 'formik';
import validateForm from '../../../Validators/Form';

class NewCustomerForm extends CustomerForm {}

const NewCustomerFormWithFormik = withFormik({
  mapPropsToValues: props => {
    return {
      customer: {
        phones: [{}],
        addresses: [{}],
        emails: [{}],
      },
    };
  },
  enableReinitialize: true,
  validate: values => {
    return {
      customer: validateForm(values.customer),
    };
  },
})(NewCustomerForm);

const NewCustomerFormWithMutation = graphql(CreateCustomerMutation)(NewCustomerFormWithFormik);

export { NewCustomerFormWithFormik as NewCustomerForm };
export default withRouter(NewCustomerFormWithMutation);
