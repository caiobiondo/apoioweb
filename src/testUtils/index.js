import React from 'react';
import { mount as enzymeMount } from 'enzyme';
import { IntlProvider } from 'react-intl';
import { locale, flattenMessages, messages } from 'locale/index';
import { ThemeProvider, theme } from 'natura-ui';

export const mount = component => {
  return enzymeMount(
    <ThemeProvider theme={theme}>
      <IntlProvider locale={locale} messages={flattenMessages(messages[locale])}>
        {component}
      </IntlProvider>
    </ThemeProvider>,
  );
};

export const clickOnElementWithText = (element, selector, text, customEvent = null) => {
  element.find(selector).forEach(toClick => {
    if (toClick.text() === text) {
      toClick.simulate('click');

      if (customEvent) toClick.simulate(customEvent);
    }
  });
};

export const changeInputValue = (form, name, value) => {
  const input = form.find(`[name="${name}"]`).get(0);
  input.props.onChange({ target: { value, name }, persist: jest.fn() });
};
