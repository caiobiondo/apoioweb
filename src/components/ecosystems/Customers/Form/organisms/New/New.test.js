import React from 'react';
import { NewCustomerForm } from './New';
import 'jest-styled-components';
import { mount, clickOnElementWithText, changeInputValue } from 'testUtils';
import { translate } from 'locale';

describe('NewCustomerForm', () => {
  const fakeCustomerData = () => {
    return {
      emails: [{ email: 'mail@test.com' }],
      addresses: [
        {
          /* eslint-disable camelcase */
          zipcode: '01219000',
          street_name: 'Rua do Arouche',
          street_number: '23',
          additional_address: 'Sexto andar',
          neighborhood: 'República',
          city: 'São Paulo',
          state: 'SP',
          /* eslint-enable camelcase */
        },
      ],
      phones: [{ phone: '(11) 1234-5678' }],
      name: 'Test Name',
      gender: 'm',
      birthday: '1943-12-30',
    };
  };

  const setup = propOverrides => {
    const fakePromise = {
      then: jest.fn(),
      catch: jest.fn(),
    };

    const mutateMock = jest.fn(() => fakePromise);

    const props = Object.assign(
      {
        history: {
          back: jest.fn(),
        },
        mutate: mutateMock,
      },
      propOverrides,
    );

    const result = mount(<NewCustomerForm {...props} />);

    return {
      props,
      result,
    };
  };

  let nextStepButtonText, submitStepButtonText;

  beforeEach(() => {
    const constantDate = new Date('2017-12-15T04:41:20');

    global.Date = class extends Date {
      constructor() {
        super();
        return constantDate;
      }
    };

    submitStepButtonText = translate('formCustomerRegister');
    nextStepButtonText = translate('formCustomerNext');
  });

  it('correctly renders the form', () => {
    const { result } = setup();

    expect(result).toMatchSnapshot();
  });

  it('correctly calls the mutate function with the customer', () => {
    const fakeCustomer = fakeCustomerData();
    const { result: form, props } = setup();
    changeInputValue(form, 'customer.name', fakeCustomer.name);
    changeInputValue(form, 'customer.gender', fakeCustomer.gender);
    changeInputValue(form, 'customer.birthday', fakeCustomer.birthday);
    changeInputValue(form, 'customer.phones.0.phone', fakeCustomer.phones[0].phone);
    changeInputValue(form, 'customer.emails.0.email', fakeCustomer.emails[0].email);

    clickOnElementWithText(form, 'button', nextStepButtonText);

    /* eslint-disable camelcase */
    changeInputValue(form, 'customer.addresses.0.zipcode', fakeCustomer.addresses[0].zipcode);
    changeInputValue(
      form,
      'customer.addresses.0.street_name',
      fakeCustomer.addresses[0].street_name,
    );
    changeInputValue(
      form,
      'customer.addresses.0.street_number',
      fakeCustomer.addresses[0].street_number,
    );
    changeInputValue(
      form,
      'customer.addresses.0.additional_address',
      fakeCustomer.addresses[0].additional_address,
    );
    changeInputValue(
      form,
      'customer.addresses.0.neighborhood',
      fakeCustomer.addresses[0].neighborhood,
    );
    changeInputValue(form, 'customer.addresses.0.city', fakeCustomer.addresses[0].city);
    changeInputValue(form, 'customer.addresses.0.state', fakeCustomer.addresses[0].state);
    /* eslint-enable camelcase */

    clickOnElementWithText(form, 'button', nextStepButtonText);
    clickOnElementWithText(form, 'button', submitStepButtonText, 'submit');

    expect(props.mutate.mock.calls.length).toBe(1);
    expect(props.mutate.mock.calls[0]).toEqual([
      {
        variables: {
          input: {
            ...fakeCustomer,
            phones: [
              {
                ...fakeCustomer.phones[0],
                phone: fakeCustomer.phones[0].phone + ' ',
              },
            ],
          },
        },
      },
    ]);
  });
});
