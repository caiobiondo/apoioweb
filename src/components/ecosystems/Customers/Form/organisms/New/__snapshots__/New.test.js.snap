// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewCustomerForm correctly renders the form 1`] = `
<Formik
  enableReinitialize={true}
  history={
    Object {
      "back": [Function],
    }
  }
  initialValues={
    Object {
      "customer": Object {
        "addresses": Array [
          Object {},
        ],
        "emails": Array [
          Object {},
        ],
        "phones": Array [
          Object {},
        ],
      },
      "editMode": false,
    }
  }
  isInitialValid={false}
  mutate={[Function]}
  onSubmit={[Function]}
  render={[Function]}
  shallow={true}
  validate={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
/>
`;
